fix kubectl --as --as-group --request-timeout inside pods

CaoShuFeng opened the pull request:
Fixes:
https://github.com/kubernetes/kubernetes/issues/49622
https://github.com/kubernetes/kubernetes/issues/49343

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```
Fix --as, --as-group, --request-timeout for kubectl running inside a pod.
```

CaoShuFeng commented:
@deads2k 
@supereagle 
@wanghaoran1988 
Please take a look.

supereagle commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -332,7 +332,9 @@ func (rules *ClientConfigLoadingRules) GetExplicitFile() string {
 
-// IsDefaultConfig returns true if the provided configuration matches the default
+// IsDefaultConfig returns true if the provided configuration matches the default, but
+// config.Timeout and config.Impersonate are omitted because user my also use --as,
```
s/my/may/

CaoShuFeng replied:
Done.

deads2k commented:
@kubernetes/sig-cli-bugs

ericchiang commented:
/ok-to-test

CaoShuFeng commented on staging/src/k8s.io/client-go/tools/clientcmd/client_config.go:
```diff
@@ -473,18 +473,25 @@ func (config *inClusterClientConfig) ClientConfig() (*restclient.Config, error)
 
-	// in-cluster configs only takes a host, token, or CA file
-	// if any of them were individually provided, ovewrite anything else
+	// in-cluster configs only takes --as, --as-group, --request-timeout
```
This is a good question.
Because overrides this value here doesn't make any sense.
When we use -s to override config.overrides.ClusterInfo.Server, we will use mergedConfig rather than inClusterConfig
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/clientcmd/merged_client_builder.go#L116
So override this value here doesn't make any sense, this logic will never run.

CaoShuFeng replied:
Removing token break e2e test here:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/kubectl/kubectl.go#L684
Override icc.Host doesn't make sense.
But override BearerToken and CAFile makes sense.
I will took all them back.

CaoShuFeng commented:
/test pull-kubernetes-e2e-kops-aws

CaoShuFeng commented:
/test pull-kubernetes-e2e-gce-etcd3

CaoShuFeng commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce-etcd3

CaoShuFeng commented:
/test pull-kubernetes-e2e-kops-aws
/test pull-kubernetes-e2e-gce-etcd3

CaoShuFeng commented:
/test pull-kubernetes-e2e-kops-aws

ericchiang commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -332,7 +332,9 @@ func (rules *ClientConfigLoadingRules) GetExplicitFile() string {
 }
 
-// IsDefaultConfig returns true if the provided configuration matches the default
+// IsDefaultConfig returns true if the provided configuration matches the default, but
```
Can you explain why doing this fixes the bug? My naive take is that doing `reflect.DeepEqual` is a really weird outside of tests, and I can't really reason when or why you'd use a method like `IsDefaultConfig`.

liggitt replied:
this is revealing that the comparison is being done at the wrong point...

currently, this is what is happening:
1. load kubeconfig if specified
2. merge in command line overrides
3. if the result is equal to what you would get if you loaded an empty kubeconfig file, discard it and switch to the in-cluster-config

The comparison should be done after step 1, then apply command line overrides. This is a more fundamental (but correct) change

CaoShuFeng replied:
> The comparison should be done after step 1, then apply command line overrides.

Hi, I tried your suggestion.
If we change the order of comparison steps, It will not make thinks less complex, I think.
For example, if user specify --namespace=foo, will we apply namespace foo to in-cluster config?

CaoShuFeng commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -332,7 +332,9 @@ func (rules *ClientConfigLoadingRules) GetExplicitFile() string {
 }
 
-// IsDefaultConfig returns true if the provided configuration matches the default
+// IsDefaultConfig returns true if the provided configuration matches the default, but
```
When user didn't specify kubeconfig file or  options like "--server", "--token", "--user", kubectl will try to read in-cluster configuration(service account).

After this change, when user specify "--as", "--as-group" or "--timeout" but doesn't specify other auth infos, IsDefaultConfig() still returns true, so that we can try to use in-cluster configuration when --as, --as-group is specified.

liggitt commented on staging/src/k8s.io/client-go/tools/clientcmd/loader.go:
```diff
@@ -341,6 +343,10 @@ func (rules *ClientConfigLoadingRules) IsDefaultConfig(config *restclient.Config
 	}
+
+	defaultConfig.Timeout = config.Timeout
+	defaultConfig.Impersonate = config.Impersonate
```
a config that specifies impersonation options is definitely *not* a default config

CaoShuFeng replied:
I dropped func IsDefaultConfig, and introuduce a new private function to check whether we should use in-cluster config.

CaoShuFeng commented:
/test pull-kubernetes-federation-e2e-gce

CaoShuFeng commented:
/test pull-kubernetes-federation-e2e-gce

liggitt commented:
> I dropped func IsDefaultConfig, and introuduce a new private function to check whether we should use in-cluster config.

do not drop exported functions from client-go... external tools depend on that

CaoShuFeng committed:
fix kubectl --as --as-group --request-timeout inside pods

Fixes:
https://github.com/kubernetes/kubernetes/issues/49622
https://github.com/kubernetes/kubernetes/issues/49343

CaoShuFeng commented:
/test pull-kubernetes-unit

CaoShuFeng commented:
/test pull-kubernetes-federation-e2e-gce
/test pull-kubernetes-e2e-gce-gpu
/test pull-kubernetes-unit

CaoShuFeng commented:
> do not drop exported functions from client-go... external tools depend on that
Done. Take it back.
